src/Moodle.ts

```diff
@@ -1,111 +1,181 @@
-            await this.refresh();
-        }
-		return !!this.cookies;
+    res = await this.fetch(this.appendToURL("/login/index.php"), {
+      headers: {
+        cookie:
+          res.headers
+            .get("set-cookie")
+            ?.split("Secure, ")
+            .find((c: String) => c.startsWith("MoodleSession")) || "",
+      },
+      method: "POST",
+      body: form,
+      redirect: "manual",
+      credentials: "include",
+    });
+    this.cookies = res.headers
+      .get("set-cookie")
+      ?.split("Secure, ")
+      .find((c: String) => c.startsWith("MoodleSession"));
+
+    if (this.cookies && options.refresh === true) {
+      await this.refresh();
     }
+    return !!this.cookies;
+  }
+
+  getCourseAndDate(anchors: Cheerio<Element>, $: CheerioAPI) {
+    const res: { date?: Date; course?: Course } = {};
+    anchors.map((_, el) => {
+      if (typeof el.attribs["href"] === "undefined") return;
+      const href = el.attribs["href"];
+      if (href?.includes("calendar")) {
+        const timeParams = href.match(/time=\d*/);
+        if (!timeParams || timeParams?.length === 0) return;
+        const timeString = timeParams[0].split("=").pop();
+        if (!timeString) return;
+        res.date = new Date(parseInt(timeString) * UnixTimestampRate);
+      }
+      if (href?.includes("course")) {
+        const courseParams = href.match(/id=\d*/);
+        if (!courseParams || courseParams?.length === 0) return;
+        const idString = courseParams[0].split("=").pop();
+        if (!idString) return;
+        res.course = {
+          id: parseInt(idString),
+          name: $(el).text() || "",
+          tasks: [],
+        };
+      }
+    });
+    return res;
+  }
+
+  /**
+   * Fetches the user data and stores them in the Moodle instance
+   * @param cookies optional
+   */
+  async refresh(cookies = this.cookies) {
+    const res = await this.fetch(
+      `${this.url}/calendar/view.php?view=upcoming`,
+      {
+        headers: { cookie: cookies || "" },
+      }
+    );
+    const body = await res.text();
+    const $ = load(body);
+    const courses = new Map<number, Course>();
+
+    try {
+      // parse courses
+      this.tasks = $(".event")
+        .map((_, el) => {
+          const description = $(".description-content", el).first().text();
+          const anchors = $("a", el);
+          const { date, course: retrievedCourse } = this.getCourseAndDate(
+            anchors,
+            $
+          );
+          if (!retrievedCourse) return;
+          if (!courses.has(retrievedCourse.id)) {
+            courses.set(retrievedCourse.id, retrievedCourse);
+          }
+          const course = courses.get(retrievedCourse.id);
+          const task = {
+            id: parseInt(el.attribs["data-event-id"]),
+            name: el.attribs["data-event-title"],
+            description,
+            deadline: date,
+            // Do we really need to crossreference like this?
```

Comment:
Thanks for pointing this out. I was pretty new to TS when I wrote this hehe.

I suppose it would be better to use the ids for referencing, yes?
```ts
export interface Task {
  // ...
  courseId: number;
}
export interface Course {
  // ...
  taskIds: Array<number>;
}

Reply:
Yes, i think it is better with the ids